### PR TITLE
Bug 2141095: [RDR] Storage System page on ACM Hub is visible even when data observability is not enabled

### PR DESCRIPTION
**Pull request by debjyoti-pandit**
Storage system page visible since acm observability is present: 

image
image

### CODE REVIEW

File: packages/mco/types/acm.ts [hunk 1]
```diff
@@ -0,0 +1,13 @@
...
+} from '@openshift-console/dynamic-plugin-sdk';
+import { K8sResourceCondition } from '../../shared/types';
+
+export type ACMMultiClusterObservability = K8sResourceCommon & {
+  status?: {
+    conditions: K8sResourceCondition[];
+  };
+};
+
+export type FeatureDetector = (setFlag: SetFeatureFlag) => Promise<void>;
```
Comment (GowthamShanmugam) [inline packages/mco/types/acm.ts:13]: Move this line inside feature.ts

File: packages/mco/features.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+export const setACMMCOFlag: FeatureDetector = async (
+  setFlag: SetFeatureFlag
+) => {
+  // This set the flag to true only when ACM MultiClusterObservability CRD is in ready state.
+  // which is later used by enable storage system.
+
+  let acmMcoIntervalId = null;
+  const acmMcoDetector = async () => {
+    try {
+      const managedClusters: ACMMultiClusterObservability[] =
```
Comment (GowthamShanmugam) [inline packages/mco/features.ts:21]:
```suggestion
      const acmMCOList: ACMMultiClusterObservability[] =
```

Comment (GowthamShanmugam) [reply packages/mco/features.ts:21]:
```
import {
  k8sList,
  K8sResourceCommon,
  SetFeatureFlag,
} from '@openshift-console/dynamic-plugin-sdk';
import { SECOND, DEFAULT_SYNC_TIME } from './constants';
import { AcmMultiClusterObservabilityModel } from './models';
import { ACMMultiClusterObservability } from './types';

type FeatureDetector = (
  setFlag: SetFeatureFlag,
  flagKey: string,
  intervalId: any
) => Promise<void>;

type FlagController = {
  flagKey: string;
  featureDetector: FeatureDetector;
  syncTime?: number;
};

const acmMcoDetector: FeatureDetector = async (
  setFlag: SetFeatureFlag,
  flagKey: string,
  intervalId: any
) => {
  // This set the flag to true only when ACM MultiClusterObservability CRD is in ready state.
  // which is later used by enable storage system.
  try {
    const managedClusters: ACMMultiClusterObservability[] =
      (await k8sList<K8sResourceCommon>({
        model: AcmMultiClusterObservabilityModel,
        queryParams: {},
      })) as ACMMultiClusterObservability[];
    const isEnabled = managedClusters?.[0].status?.conditions?.some(
      (condition) => condition.status === 'True' && condition.type === 'Ready'
    );
    if (isEnabled) {
      setFlag(flagKey, true);
      clearInterval(intervalId);
    }
  } catch (error) {
    setFlag(flagKey, false);
  }
};

const flagController: FlagController[] = [
  {
    flagKey: 'ACM_MCO',
    featureDetector: acmMcoDetector,
    syncTime: 60,
  },
];

export const setFeatureFlag = async (setFlag: SetFeatureFlag): Promise<void> => {
  flagController.map(async (controller) => {
    let intervalId = null;
    await controller.featureDetector(setFlag, controller.flagKey, intervalId);
    intervalId = setInterval(
      controller.featureDetector,
      (controller.syncTime || DEFAULT_SYNC_TIME) * SECOND,
      setFlag, controller.flagKey, intervalId
    );
  });
};

```

What do you think about this change?
No dont need to call each flag set function from consoleExtension.json
just calling the below function will set all flags
```
{
    "type": "console.flag",
    "properties": {
      "handler": {
        "$codeRef": "features.setFeatureFlag"
      }
    }
  },
```

Comment (GowthamShanmugam) [reply packages/mco/features.ts:21]: cc: @SanjalKatiyar any thoughts?

Comment (GowthamShanmugam) [reply packages/mco/features.ts:21]: it won't be null once the setInterval is called intervalId gets populated. When controller.featureDetector is called the intervalId reference is passed.

Comment (SanjalKatiyar) [reply packages/mco/features.ts:21]: it is primitive type "number"

Comment (SanjalKatiyar) [reply packages/mco/features.ts:21]: it is passed by value

Comment (GowthamShanmugam) [reply packages/mco/features.ts:21]: let intervals: number[] = []

```
export const setFeatureFlag = async (setFlag: SetFeatureFlag): Promise<void> => {
  flagController.map(async (controller, id) => {
    await controller.featureDetector(setFlag, controller.flagKey, id);
    const intervalId = setInterval(
      controller.featureDetector,
      (controller.syncTime || DEFAULT_SYNC_TIME) * SECOND,
      setFlag, controller.flagKey, intervalId
    );
    intervels[id] = intervalId
  });
};

```

inside acmMcoDetector function
clearInterval(intervals[id]);

Comment (SanjalKatiyar) [reply packages/mco/features.ts:21]: this should work !! looks good...

**Comment by SanjalKatiyar** [comment]
/retitle Bug 2141095: [RDR] Storage System page on ACM Hub is visible even when data observability is not enabled

File: packages/mco/features.ts (hunk 2 above)
Comment (GowthamShanmugam) [reply packages/mco/features.ts:21]: @debjyoti-pandit Lets switch to this approach

Comment (debjyoti-pandit) [reply packages/mco/features.ts:21]: ack

File: packages/mco/features.ts [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+    featureDetector: acmMcoDetector,
+  },
+];
+
+export const setFeatureFlag = async (
+  setFlag: SetFeatureFlag
+): Promise<void> => {
+  flagController.map(async (controller, id) => {
+    await controller.featureDetector(setFlag, controller.flagKey, id);
+    const intervalId = setInterval(
```
Comment (GowthamShanmugam) [inline packages/mco/features.ts:61]: This wont work, we already discussed about this in a comment and modified the logic, please check

File: packages/mco/features.ts [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    setFlag(flagKey, false);
+  }
+};
+
+const flagController: FlagController[] = [
+  {
+    flagKey: 'ACM_MCO',
+    featureDetector: acmMcoDetector,
+  },
+];
```
[the remark this answers is not in the document]
Comment (debjyoti-pandit) [reply packages/mco/features.ts:54]: ack

**Review by GowthamShanmugam — APPROVED** [review]

**Comment by debjyoti-pandit** [comment]
Before: 
image

**Pull request merged**